embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/CustomBreadcrumbDataSource.kt

```diff
@@ -1,32 +1,41 @@
 package io.embrace.android.embracesdk.capture.crumbs
 
-import android.text.TextUtils
-import io.embrace.android.embracesdk.arch.DataCaptureService
+import io.embrace.android.embracesdk.arch.datasource.DataSourceImpl
+import io.embrace.android.embracesdk.arch.destination.SessionSpanWriter
+import io.embrace.android.embracesdk.arch.destination.SpanEventData
+import io.embrace.android.embracesdk.arch.destination.SpanEventMapper
+import io.embrace.android.embracesdk.arch.limits.UpToLimitStrategy
 import io.embrace.android.embracesdk.config.ConfigService
-import io.embrace.android.embracesdk.logging.InternalEmbraceLogger
-import io.embrace.android.embracesdk.logging.InternalStaticEmbraceLogger
+import io.embrace.android.embracesdk.internal.clock.millisToNanos
 import io.embrace.android.embracesdk.payload.CustomBreadcrumb
 
 /**
  * Captures custom breadcrumbs.
  */
 internal class CustomBreadcrumbDataSource(
-    private val configService: ConfigService,
-    private val store: BreadcrumbDataStore<CustomBreadcrumb> = BreadcrumbDataStore {
-        configService.breadcrumbBehavior.getCustomBreadcrumbLimit()
-    },
-    private val logger: InternalEmbraceLogger = InternalStaticEmbraceLogger.logger
-) : DataCaptureService<List<CustomBreadcrumb>> by store {
+    configService: ConfigService,
+    writer: SessionSpanWriter
+) : DataSourceImpl<SessionSpanWriter>(
+    destination = writer,
+    limitStrategy = UpToLimitStrategy(configService.breadcrumbBehavior::getCustomBreadcrumbLimit)
+),
+    SpanEventMapper<CustomBreadcrumb> {
 
     fun logCustom(message: String, timestamp: Long) {
-        if (TextUtils.isEmpty(message)) {
-            logger.logWarning("Breadcrumb message must not be blank")
-            return
-        }
-        try {
-            store.tryAddBreadcrumb(CustomBreadcrumb(message, timestamp))
-        } catch (ex: Exception) {
-            logger.logError("Failed to log custom breadcrumb with message $message", ex)
-        }
+        captureData(
+            inputValidation = {
+                message.isNotEmpty()
+            },
+            captureAction = {
+                val crumb = CustomBreadcrumb(message, timestamp)
+                addEvent(crumb.toSpanEventData())
```

Comment:
I wonder if we can make the `toSpanEventData` call in the `addEvent` method - so the datasource will just need to pass in a breadcrumb that can implicitly convert itself and have that conversion down at a lower layer. 

Are we able to create an interface that covers that so all POJOs that will end up becoming a SpanEvent will just implement that?

Reply:
@bidetofevil our hands are tied a little here as the POJOs will be generated from the OpenAPI schema & it's non-trivial to have them directly implement an interface like that. I have however refactored the code so that a mapper function can be passed in, which is probably the nearest we can get.